# 1. What is the current pattern of herbicide resistances across SE Australia?

## 1.5. Models of herbicide resistances using population-level genomic information

Can we do even better in terms of predicting herbicide resistances across SE Australia using genomic information compared with simply using the geographic coordinates from which the populations were sampled from?

## GWAS-based: identify putative herbicide resistance QTL, finetune the estiamted effects, and predict using these QTL

```R
source("src/modelling_and_crossfold_validation.r")

SUMMARISE_AND_PLOT = function(fname_genotype="res/genotype_data-allele_frequencies-p_minus_one.csv", 
                              fname_phenotype="res/phenotype_data.csv",
                              phenotype_name="Glyphosate",
                              batch="all",
                              r=5,
                              k=10,
                              maf=0.01) {
    # fname_genotype = "res/genotype_data-allele_frequencies-p_minus_one.csv"
    # fname_phenotype = "res/phenotype_data.csv"
    # phenotype_name = "Glyphosate"
    # batch = "all"
    # r = 5
    # k = 10
    # maf = 0.05

    variable_names = c("X.Population", "Pool_size", "Latitude", "Longitude", phenotype_name)
    
    df_x_tmp = read.csv(fname_genotype)
    df_x = data.frame(X.Population=colnames(df_x_tmp)[4:ncol(df_x_tmp)], t(df_x_tmp[, 4:ncol(df_x_tmp), ]))
    colnames(df_x) = c("X.Population", paste0(df_x_tmp$X.chr, "-", df_x_tmp$pos, "-", df_x_tmp$allele))
        
    df_y = read.csv(fname_phenotype)
    if (batch!="all") {
        idx_rows = df_y$Batch == batch
        if (sum(idx_rows) == 0) {
            print("@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@")
            print("ERROR!")
            print("Batch ID to include is incorrect. Please select from 2018 and 2019.")
            return("ERROR!")
        }
        df_y = df_y[idx_rows, ]
    }
    idx_phen_col = which(
        (colnames(df_y) == variable_names[1]) |
        (colnames(df_y) == variable_names[2]) |
        (colnames(df_y) == variable_names[3]) |
        (colnames(df_y) == variable_names[4]) |
        (colnames(df_y) == phenotype_name))
    df_y = df_y[, idx_phen_col]
    df_y = df_y[!is.na(df_y[, colnames(df_y) == phenotype_name]), ]
    
    DATA = merge(df_y, df_x, by=variable_names[1])
    idx_y_col = which(colnames(DATA) == phenotype_name)
    idx_x_col = which(!((colnames(DATA) == variable_names[1]) |
                        (colnames(DATA) == variable_names[2]) |
                        (colnames(DATA) == variable_names[3]) |
                        (colnames(DATA) == variable_names[4]) |
                        (colnames(DATA) == phenotype_name)))
    y = as.matrix(DATA[, idx_y_col])
    x = as.matrix(DATA[, idx_x_col])
    x = x[, !((colMeans(x) < maf) | (colMeans(x) > (1-maf)))]

    print(paste0("n = ", nrow(x)))
    print(paste0("p = ", ncol(x)))
    print(paste0("k-fold = ", k))
    print(paste0("replications = ", r))


    ### K-fold cross-validation
    options(digits.secs=7)
    start_time = Sys.time()
    kfold_out = KFOLD_CV(x, y, r, k)
    end_time = Sys.time()
    print(end_time - start_time)

    ### Summarise per model per individual
    models = names(kfold_out)
    df_ys = data.frame()
    df_summary = data.frame()
    for (i in 1:length(models)) {
        rep = c()
        pop = c()
        y_test = c()
        y_hat = c()
        for (j in 1:r) {
            for (l in 1:k) {
                # i = 1; j = 1; l = 1
                dat = eval(parse(text=paste0("kfold_out$", models[i], "$rep_", j, "$fold_", l)))
                if (length(dat) > 0) {
                    rep = c(rep, rep(j, length(dat$idx_test)))
                    pop = c(pop, dat$idx_test)
                    y_test = c(y_test, dat$y_test)
                    y_hat = c(y_hat, dat$y_hat)
                }
            }
        }
        df = data.frame(rep, pop, y_test, y_hat)
        df = df[order(df$pop), ]
        ### Dataframe for plotting the prediction statistics for each population (df_ys)
        tru = aggregate(y_test ~ pop, data=df, FUN=mean); colnames(tru) = c("pop", "y_true")
        mus = aggregate(y_hat ~ pop, data=df, FUN=mean); colnames(mus) = c("pop", "y_pred_mean")
        sds = aggregate(y_hat ~ pop, data=df, FUN=sd); colnames(sds) = c("pop", "y_pred_sd")
        merged = merge(merge(tru, mus, by="pop"), sds, by="pop")
        if (nrow(df_ys)== 0) {
            df_ys = cbind(model=rep(models[i], nrow(merged)), merged)
        } else {
            df_ys = rbind(df_ys, cbind(model=rep(models[i], nrow(merged)), merged))
        }
        ### Dataframe for assessing each model
        corr = c()
        rmse = c()
        for (j in 1:r) {
            # j = 1
            dat = df[df$rep==j, ]
            corr = c(corr, tryCatch(cor(dat$y_test, dat$y_hat), warning=function(e){0.0}))
            rmse = c(rmse, tryCatch(sqrt(mean((dat$y_test-dat$y_hat)^2)), warning=function(e){0.0}))
        }
        perf = data.frame(model=models[i], corr=mean(corr, na.rm=TRUE), rmse=mean(rmse, na.rm=TRUE), corr_sd=sd(corr, na.rm=TRUE), rmse_sd=sd(rmse, na.rm=TRUE))
        if (nrow(df_summary) == 0) {
            df_summary = perf
        } else {
            df_summary = rbind(df_summary, perf)
        }
    }

    for (i in 1:length(models)) {    
        # i = 1
        df_ys_sub = df_ys[df_ys$model==models[i], ]
        corr = df_summary$corr[df_summary$model==models[i]]
        rmse = df_summary$rmse[df_summary$model==models[i]]
        
        svg(paste0("genotype_to_phenotype_prediction---", models[i], "-", phenotype_name, "-gp.svg"))
        plot(x=range(df_ys_sub$y_true),
             y=range(df_ys_sub$y_true),
             xlab="Observed",
             ylab="Predicted",
             main=paste0(phenotype_name, " resistance (Model: ", models[i], ")"),
             type="n")
        grid()
        for (j in 1:nrow(df_ys_sub)) {
            points(x=df_ys_sub$y_true[j], y=df_ys_sub$y_pred_mean[j], pch=19)
            arrows(x0=df_ys_sub$y_true[j],
                   x1=df_ys_sub$y_true[j],
                   y0=df_ys_sub$y_pred_mean[j]-df_ys_sub$y_pred_sd[j],
                   y1=df_ys_sub$y_pred_mean[j]+df_ys_sub$y_pred_sd[j],
                   code=3,
                   length=0.1,
                   angle=90,
                   col="blue")
        }
        lmod = lm(y_pred_mean ~ y_true, data=df_ys_sub)
        lmod = tryCatch(abline(lmod, col="red"),
                 error=function(e){
                    y___ = df_ys_sub$y_true+(rnorm(nrow(df_ys_sub))*1e-9)
                    lm(df_ys_sub$y_pred_mean ~ y___)
                 })
        abline(lmod, col="red")
        legend("topright", legend=c(paste0("n = ", nrow(x)),
                                    paste0("p = ", ncol(x)),
                                    paste0("k-fold = ", k),
                                    paste0("replications = ", r),
                                    paste0("corr = ", round(corr*100, 2), "%"),
                                    paste0("rmse = ", round(rmse*100/diff(range(df_ys_sub$y_true)),2), "%")
                                   )
              )
        dev.off()
    }
    return(list(df_ys=df_ys, df_summary=df_summary))
}

herbicide_resistances = c("Tri_Allate",
                          "Glyphosate",
                          "Prosulfocarb_and_S_metolachlor",
                          "Clethodim",
                          "Pyroxasulfone",
                          "Imazamox_and_Imazapyr",
                          "Bixlozone",
                          "Atrazine",
                          "Cinmethylin",
                          "Paraquat",
                          "Sulfometuron",
                          "Terbuthylazine")
for (phenotype_name in herbicide_resistances) {
    # phenotype_name = herbicide_resistances[2]
    print("@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@")
    print(phenotype_name)
    eval(parse(text=paste0(phenotype_name, " = SUMMARISE_AND_PLOT(phenotype_name='", phenotype_name, "')")))
}
```

![test](../res/genotype_to_phenotype_prediction---ols-Glyphosate-gp.svg)
![test](../res/genotype_to_phenotype_prediction---ridge-Glyphosate-gp.svg)
![test](../res/genotype_to_phenotype_prediction---lasso-Glyphosate-gp.svg)
![test](../res/genotype_to_phenotype_prediction---elastic-Glyphosate-gp.svg)

![test](../res/genotype_to_phenotype_prediction---ols-Sulfometuron-gp.svg)
![test](../res/genotype_to_phenotype_prediction---ridge-Sulfometuron-gp.svg)
![test](../res/genotype_to_phenotype_prediction---lasso-Sulfometuron-gp.svg)
![test](../res/genotype_to_phenotype_prediction---elastic-Sulfometuron-gp.svg)

![test](../res/genotype_to_phenotype_prediction---ols-Terbuthylazine-gp.svg)
![test](../res/genotype_to_phenotype_prediction---ridge-Terbuthylazine-gp.svg)
![test](../res/genotype_to_phenotype_prediction---lasso-Terbuthylazine-gp.svg)
![test](../res/genotype_to_phenotype_prediction---elastic-Terbuthylazine-gp.svg)

## GP using genomewide allele frequency information

## GP using kindhip information

## Findings
