# 1. What is the current pattern of herbicide resistances across SE Australia?

## 1.6. Models of herbicide resistances across SE Austrlia using population-level genomic information, environmental variables, and other herbicide resistances

Add these data from CSIRO (c/o Alex): [csv file](https://urldefense.com/v3/__https://drive.google.com/file/d/1RDxAuoXQZynMORNE2kRP_ZqHkWx1FuW4/view?usp=drive_link__;!!C5rN6bSF!EW1RdWcKYKRux59Bb7iOcZFy6tcDY7A5eX9aNZsdb1sGhy_ACkbt8TvfAgDGqldsbFJIl4y-NOBX5svFGLNcdxOytNbQJto_WFOeesgEtPsMrYp6$)

```R
setwd("/data-weedomics-1/weedomics/exploratory")
source("src/modelling_and_crossfold_validation.r")

fname_phenotype = "res/phenotype_data.csv"
fname_genotype = "res/genotype_data-allele_frequencies-p_minus_one.csv"
dname_climate_grid_data = "res/bom_grid_data_1991_to_2020"
dname_soil_grid_data = "res/fao_soil_data_version1.2"
fname_csiro_soil_data = "res/CSIRO_AusSoil_complete.csv"

df_genotype = LOAD_GENOTYPES(fname_genotype=fname_genotype, maf=0.01)
raster_layers = LOAD_ENVIRONMENTAL_DATA(dname_climate_grid_data=dname_climate_grid_data,
                                dname_soil_grid_data=dname_soil_grid_data,
                                extension_name="asc")
df_covariates_csiro_soil_data = read.csv(file=fname_csiro_soil_data)
colnames(df_covariates_csiro_soil_data)[grepl("Accession", colnames(df_covariates_csiro_soil_data))] = "X.Population"

vec_names_ids = c("X.Population", "Batch", "Pool_size", "Latitude", "Longitude", "long", "lat")
vec_names_herbicides = colnames(read.csv(fname_phenotype))
vec_names_herbicides = vec_names_herbicides[!(vec_names_herbicides %in% vec_names_ids)]

vec_models = c("ols", "ridge", "lasso", "elastic", "bayesa", "bayesb", "bayesc", "logistic_ridge", "logistic_lasso", "logistic_elastic")
vec_explanatory_variables = c("E", "G", "E_and_G")
logistic_y_bin_threshold = 0.2 ### 20% resistance

for (herbi in vec_names_herbicides) {
    # herbi = vec_names_herbicides[2]
    print("########################")
    print(herbi)
    df_phenotype = LOAD_PHENOTYPES(fname_phenotype=fname_phenotype, batch="all", phenotype_names=herbi)
    df_phenotype = merge(df_phenotype, df_covariates_csiro_soil_data, by="X.Population")
    df = MERGE_PHENOTYPE_WITH_GENOTYPE_AND_ENVIRONMENTAL_DATA(df_phenotype=df_phenotype, df_genotype=df_genotype, raster_layers=raster_layers)
    vec_names_csiro_soil_data = colnames(df_covariates_csiro_soil_data)
    vec_names_csiro_soil_data = vec_names_csiro_soil_data[!(vec_names_csiro_soil_data %in% vec_names_ids)]
    vec_names_phenotypes = colnames(df_phenotype)
    vec_names_phenotypes = vec_names_phenotypes[!(vec_names_phenotypes %in% vec_names_ids)]
    vec_names_phenotypes = vec_names_phenotypes[!(vec_names_phenotypes %in% vec_names_csiro_soil_data)]
    vec_names_loci = colnames(df_genotype)
    vec_names_loci = vec_names_loci[!(vec_names_loci %in% vec_names_ids)]
    # vec_names_fao_bom_data = colnames(df)
    # vec_names_fao_bom_data = vec_names_fao_bom_data[!(vec_names_fao_bom_data %in% c(vec_names_ids, vec_names_csiro_soil_data, vec_names_phenotypes, vec_names_loci))]
    y = df[, colnames(df) %in% herbi]
    G = df[, colnames(df) %in% vec_names_loci]
    S = df[, colnames(df) %in% vec_names_csiro_soil_data]

    if (var(y, na.rm=TRUE) < 0) {
        print("No variance in the phenotype data.")
        next
    }
    vec_var_loci = apply(G, MARGIN=2, FUN=var, na.rm=TRUE)
    G = G[, vec_var_loci != 0]
    vec_var_csiro_soil_data = apply(S, MARGIN=2, FUN=var)
    S = S[, !((vec_var_csiro_soil_data == 0) | (is.na(vec_var_csiro_soil_data)))]

    y = scale(y, scale=TRUE, center=TRUE)
    G = scale(G, scale=TRUE, center=TRUE)
    S = scale(S, scale=TRUE, center=TRUE)

    for (explanatory_variables in vec_explanatory_variables) {
        # explanatory_variables = "E"
        print(explanatory_variables)
        options(digits.secs=7)
        start_time = Sys.time()
        if (explanatory_variables == "E") {
            eval(parse(text=paste0("kfold_", explanatory_variables, " = KFOLD_CV(x=S, y=y, r=10, k=10, vec_models=vec_models, logistic_y_bin_threshold=logistic_y_bin_threshold)")))
        } else if (explanatory_variables == "G") {
            eval(parse(text=paste0("kfold_", explanatory_variables, " = KFOLD_CV(x=G, y=y, r=10, k=10, vec_models=vec_models, logistic_y_bin_threshold=logistic_y_bin_threshold)")))
        } else if (explanatory_variables == "E_and_G") {
            eval(parse(text=paste0("kfold_", explanatory_variables, " = KFOLD_CV(x=cbind(S, G), y=y, r=10, k=10, vec_models=vec_models, logistic_y_bin_threshold=logistic_y_bin_threshold)")))
        }
        end_time = Sys.time()
        print(end_time - start_time)
    }

    ### Plot per model which contains predictions of using genomic data only, and genomic data with CSIRO soil data
    col_bars = rgb(0.1, 0.1, 0.1, alpha=0.5)
    col_points = rgb(0.9, 0.2, 0.1, alpha=0.5)
    col_line = rgb(0.9, 0.2, 0.1, alpha=0.5)
    for (model in vec_models) {
        # model = "ols"
        # model = "logistic_lasso"
        svg(paste0("res/genomic_prediction-", model, "-", herbi, ".svg"), width=15.75, height=5.25)
        layout(matrix(1:3, nrow=1))
        for (explanatory_variables in vec_explanatory_variables) {
            # explanatory_variables = "E"
            ### Extract expected and predicted response variable
            y_idx = suppressWarnings(eval(parse(text=paste0("unlist(lapply(kfold_", explanatory_variables, "$", model, ", FUN=function(x){lapply(x, FUN=function(y){if(!is.na(y)){y$idx_test}})}))"))))
            y_true = suppressWarnings(eval(parse(text=paste0("unlist(lapply(kfold_", explanatory_variables, "$", model, ", FUN=function(x){lapply(x, FUN=function(y){if(!is.na(y)){y$y_test}})}))"))))
            y_pred = suppressWarnings(eval(parse(text=paste0("unlist(lapply(kfold_", explanatory_variables, "$", model, ", FUN=function(x){lapply(x, FUN=function(y){if(!is.na(y)){y$y_hat}})}))"))))
            if (grepl("logistic", model) == FALSE) {
                y_true = (y_true * attr(y, "scaled:scale")) + attr(y, "scaled:center")
                y_pred = (y_pred * attr(y, "scaled:scale")) + attr(y, "scaled:center")
            } else {
                ### Predicted binary response
                y_hat = y_pred 
                ### Linear predictors
                l_hat = suppressWarnings(eval(parse(text=paste0("unlist(lapply(kfold_", explanatory_variables, "$", model, ", FUN=function(x){lapply(x, FUN=function(y){if(!is.na(y)){y$l_hat}})}))"))))
            }
            ### Calculate prediction accuracy metrics
            if (grepl("logistic", model) == FALSE) {
                df_y = data.frame(y_idx, y_true, y_pred)
                y_true = aggregate(y_true ~ y_idx, data=df_y, FUN=mean)
                y_pred = aggregate(y_pred ~ y_idx, data=df_y, FUN=mean)
                y_pred_sd = aggregate(y_pred ~ y_idx, data=df_y, FUN=sd)
                df_y = merge(y_true, y_pred, by="y_idx"); colnames(df_y) = c("y_idx", "y_true", "y_pred_mean")
                df_y = merge(df_y, y_pred_sd, by="y_idx"); colnames(df_y) = c("y_idx", "y_true", "y_pred_mean", "y_pred_sd")
                corr = cor(df_y$y_true, df_y$y_pred_mean)
                mae = mean(abs(df_y$y_true - df_y$y_pred_mean))
                rmse = sqrt(mean((df_y$y_true - df_y$y_pred_mean)^2))
                r2 = 1.00 - (sum((df_y$y_true - df_y$y_pred_mean)^2) / sum((df_y$y_true - mean(df_y$y_true))^2))
            } else {
                logistic_accuracy = mean(y_true == y_pred)
            }
            ### Plot
            if (grepl("logistic", model) == FALSE) {
                plot(x=0, y=0, xlim=range(df_y$y_true), ylim=range(c(df_y$y_pred_mean-df_y$y_pred_sd, df_y$y_pred_mean+df_y$y_pred_sd)), type="n", 
                    xlab="Observed (survival rate, %)", 
                    ylab="Predicted (survival rate, %)", 
                    main=paste0(herbi, " Resistance\n(", model, "; y~", explanatory_variables, ")"))
                grid()
                arrows(x0=df_y$y_true, x1=df_y$y_true, y0=df_y$y_pred_mean-df_y$y_pred_sd, y1=df_y$y_pred_mean+df_y$y_pred_sd, code=3, col=col_bars, angle=90, length=0.0, lty=1)
                points(x=df_y$y_true, y=df_y$y_pred_mean, pch=19, col=col_points)
                abline(lm(y_pred_mean ~ y_true, data=df_y), lty=2, col=col_line)
                legend("bottomright", legend=c(
                    paste0("Pearson's correlation = ", round(100*corr, 2), "%"),
                    paste0("Mean absolute error = ", round(mae, 2)),
                    paste0("Root mean squared error = ", round(rmse, 2)),
                    paste0("Coefficient of determination = ", round(100*r2, 2), "%")
                ), bty="n")
            } else {
                plot(x=range(l_hat), y=c(0,1), type="n", 
                     xlab="Logit Predictor",
                     ylab="Predicted Survival", 
                    main=paste0(herbi, " Resistance\n(", model, "; Bin(y; ", round(logistic_y_bin_threshold, 2), ")~", explanatory_variables, ")"))
                points(x=l_hat, y=y_hat, col=col_points)
                mod_logit = glm(y_hat ~ l_hat, family=binomial(link="logit"))
                curve(predict(mod_logit, data.frame(l_hat=x),type='resp'), add=TRUE, lty=2, col=col_line)
                legend("topleft", 
                       legend=paste0("Prediction accuracy (concordance) = ", round(100*logistic_accuracy, 2), "%"), 
                       bty="n")
            }
        }
        dev.off()
    }
}
```

![ols-Glyphosate](../res/genomic_prediction-ols-Glyphosate.svg)
![ridge-Glyphosate](../res/genomic_prediction-ridge-Glyphosate.svg)
![lasso-Glyphosate](../res/genomic_prediction-lasso-Glyphosate.svg)
![elastic-Glyphosate](../res/genomic_prediction-elastic-Glyphosate.svg)
![bayesa-Glyphosate](../res/genomic_prediction-bayesa-Glyphosate.svg)
![bayesb-Glyphosate](../res/genomic_prediction-bayesb-Glyphosate.svg)
![bayesc-Glyphosate](../res/genomic_prediction-bayesc-Glyphosate.svg)

![ols-Clethodim](../res/genomic_prediction-ols-Clethodim.svg)
![ridge-Clethodim](../res/genomic_prediction-ridge-Clethodim.svg)
![lasso-Clethodim](../res/genomic_prediction-lasso-Clethodim.svg)
![elastic-Clethodim](../res/genomic_prediction-elastic-Clethodim.svg)
![bayesa-Clethodim](../res/genomic_prediction-bayesa-Clethodim.svg)
![bayesb-Clethodim](../res/genomic_prediction-bayesb-Clethodim.svg)
![bayesc-Clethodim](../res/genomic_prediction-bayesc-Clethodim.svg)

![ols-Paraquat](../res/genomic_prediction-ols-Paraquat.svg)
![ridge-Paraquat](../res/genomic_prediction-ridge-Paraquat.svg)
![lasso-Paraquat](../res/genomic_prediction-lasso-Paraquat.svg)
![elastic-Paraquat](../res/genomic_prediction-elastic-Paraquat.svg)
![bayesa-Paraquat](../res/genomic_prediction-bayesa-Paraquat.svg)
![bayesb-Paraquat](../res/genomic_prediction-bayesb-Paraquat.svg)
![bayesc-Paraquat](../res/genomic_prediction-bayesc-Paraquat.svg)

![ols-Atrazine](../res/genomic_prediction-ols-Atrazine.svg)
![ridge-Atrazine](../res/genomic_prediction-ridge-Atrazine.svg)
![lasso-Atrazine](../res/genomic_prediction-lasso-Atrazine.svg)
![elastic-Atrazine](../res/genomic_prediction-elastic-Atrazine.svg)
![bayesa-Atrazine](../res/genomic_prediction-bayesa-Atrazine.svg)
![bayesb-Atrazine](../res/genomic_prediction-bayesb-Atrazine.svg)
![bayesc-Atrazine](../res/genomic_prediction-bayesc-Atrazine.svg)

![ols-Prosulfocarb_and_S_metolachlor](../res/genomic_prediction-ols-Prosulfocarb_and_S_metolachlor.svg)
![ridge-Prosulfocarb_and_S_metolachlor](../res/genomic_prediction-ridge-Prosulfocarb_and_S_metolachlor.svg)
![lasso-Prosulfocarb_and_S_metolachlor](../res/genomic_prediction-lasso-Prosulfocarb_and_S_metolachlor.svg)
![elastic-Prosulfocarb_and_S_metolachlor](../res/genomic_prediction-elastic-Prosulfocarb_and_S_metolachlor.svg)
![bayesa-Prosulfocarb_and_S_metolachlor](../res/genomic_prediction-bayesa-Prosulfocarb_and_S_metolachlor.svg)
![bayesb-Prosulfocarb_and_S_metolachlor](../res/genomic_prediction-bayesb-Prosulfocarb_and_S_metolachlor.svg)
![bayesc-Prosulfocarb_and_S_metolachlor](../res/genomic_prediction-bayesc-Prosulfocarb_and_S_metolachlor.svg)

![ols-Tri_Allate](../res/genomic_prediction-ols-Tri_Allate.svg)
![ridge-Tri_Allate](../res/genomic_prediction-ridge-Tri_Allate.svg)
![lasso-Tri_Allate](../res/genomic_prediction-lasso-Tri_Allate.svg)
![elastic-Tri_Allate](../res/genomic_prediction-elastic-Tri_Allate.svg)
![bayesa-Tri_Allate](../res/genomic_prediction-bayesa-Tri_Allate.svg)
![bayesb-Tri_Allate](../res/genomic_prediction-bayesb-Tri_Allate.svg)
![bayesc-Tri_Allate](../res/genomic_prediction-bayesc-Tri_Allate.svg)
