# 1. What is the current pattern of herbicide resistances across SE Australia?

## 1.6. Models of herbicide resistances across SE Austrlia using population-level genomic information, environmental variables, and other herbicide resistances

```R
setwd("/data-weedomics-1/Lolium_rigidum_population_and_quantitative_genetics")
source("src/modelling_and_crossfold_validation.r")

df_phenotype = LOAD_PHENOTYPES(fname_phenotype="res/phenotype_data.csv",
                                batch="all",
                                phenotype_names=c("Glyphosate", "Clethodim"))
df_genotype = LOAD_GENOTYPES(fname_genotype="res/genotype_data-allele_frequencies-p_minus_one.csv",
                                maf=0.01)
raster_layers = LOAD_ENVIRONMENTAL_DATA(dname_climate_grid_data="res/bom_grid_data_1991_to_2020",
                                dname_soil_grid_data="res/fao_soil_data_version1.2",
                                extension_name="asc")
df = MERGE_PHENOTYPE_WITH_GENOTYPE_AND_ENVIRONMENTAL_DATA(df_phenotype=df_phenotype,
                                df_genotype=df_genotype,
                                raster_layers=raster_layers)

n = nrow(df)
p = ncol(df)

y       = scale(as.matrix(df[, 3:3]),                            center=F, scale=F)
# y       = scale(as.matrix(df[, 4:4]),                            center=F, scale=F)
y_covar = scale(as.matrix(df[, 4:4]),                            center=F, scale=F)
# y_covar = scale(as.matrix(df[, 3:3]),                            center=F, scale=F)
coor    = scale(as.matrix(df[, 5:6]),                            center=F, scale=F)
G       = scale(as.matrix(df[, 7:(p-length(raster_layers))]),    center=F, scale=F)
E       = scale(as.matrix(df[, (p-length(raster_layers)+1): p]), center=F, scale=F)

options(digits.secs=7)
start_time = Sys.time()
# kfold = KFOLD_CV(x=cbind(y_covar, coor, coor[,1]*coor[,2], E, G), y=y, r=10, k=10)
kfold = KFOLD_CV(x=cbind(coor, coor[,1]*coor[,2], E, G), y=y, r=10, k=10)
end_time = Sys.time()
print(end_time - start_time)

ols_corr = mean(unlist(lapply(kfold$ols, FUN=function(x){lapply(x, FUN=function(y){y$cor})})))
elastic_corr = mean(unlist(lapply(kfold$elastic, FUN=function(x){lapply(x, FUN=function(y){y$cor})})))
lasso_corr = mean(unlist(lapply(kfold$lasso, FUN=function(x){lapply(x, FUN=function(y){y$cor})})))
ridge_corr = mean(unlist(lapply(kfold$ridge, FUN=function(x){lapply(x, FUN=function(y){y$cor})})))

ols_rmse = mean(unlist(lapply(kfold$ols, FUN=function(x){lapply(x, FUN=function(y){y$rmse})})))
elastic_rmse = mean(unlist(lapply(kfold$elastic, FUN=function(x){lapply(x, FUN=function(y){y$rmse})})))
lasso_rmse = mean(unlist(lapply(kfold$lasso, FUN=function(x){lapply(x, FUN=function(y){y$rmse})})))
ridge_rmse = mean(unlist(lapply(kfold$ridge, FUN=function(x){lapply(x, FUN=function(y){y$rmse})})))

GLYPH_2 = data.frame(model=c("ols", "elastic", "lasso", "ridge"),
           corr=c(ols_corr, elastic_corr, lasso_corr, ridge_corr),
           rmse=c(ols_rmse, elastic_rmse, lasso_rmse, ridge_rmse))


y_idx = unlist(lapply(kfold$ols, FUN=function(x){lapply(x, FUN=function(y){y$idx_test})}))
y_true = unlist(lapply(kfold$ols, FUN=function(x){lapply(x, FUN=function(y){y$y_test})}))
y_pred = unlist(lapply(kfold$ols, FUN=function(x){lapply(x, FUN=function(y){y$y_hat})}))
df_y = data.frame(y_idx, y_true, y_pred)

y_true = aggregate(y_true ~ y_idx, data=df_y, FUN=mean)
y_pred = aggregate(y_pred ~ y_idx, data=df_y, FUN=mean)
y_pred_sd = aggregate(y_pred ~ y_idx, data=df_y, FUN=sd)
df_y = merge(y_true, y_pred, by="y_idx"); colnames(df_y) = c("y_idx", "y_true", "y_pred_mean")
df_y = merge(df_y, y_pred_sd, by="y_idx"); colnames(df_y) = c("y_idx", "y_true", "y_pred_mean", "y_pred_sd")




##################
### test GWAS: ###
##################
K = G %*% t(G)
PCA = prcomp(K)
X = cbind(rep(1, nrow(G)), PCA$x[,1], G)
b = t(X) %*% solve(X %*% t(X)) %*% y

# epsilon = y - (X %*% b)
# ve = sum(epsilon * epsilon) / diff(dim(X))
# vb = ve * t(X) %*% solve(X %*% t(X)) %*% solve(X %*% t(X)) %*% X

idx = 3:length(b)
pval = 2 * pnorm(abs(b[idx]), mean=mean(b[idx]), sd=sd(b[idx]), lower.tail=FALSE)
lod = -log10(pval)

P = data.frame(locus=rownames(b)[idx], b[idx], pval, lod)
P$chr = as.factor(unlist(lapply(strsplit(P$locus, "-"), FUN=function(x){x[1]})))
P$pos = as.numeric(unlist(lapply(strsplit(P$locus, "-"), FUN=function(x){x[2]})))
P$allele = unlist(lapply(strsplit(P$locus, "-"), FUN=function(x){x[3]}))
P$freq = colMeans(X[, idx])
P$consecutive_pos = NA
P$colours_chr = NA



P = P[order(P$allele), ]
P = P[order(P$pos), ]
P = P[order(P$chr), ]


vec_chr = sort(unique(P$chr))
vec_col = rep(rainbow(7), times=ceiling(length(vec_chr)/7))

minimum = min(P$pos[P$chr == vec_chr[1]], na.rm=TRUE)
P$consecutive_pos[P$chr == vec_chr[1]] = P$pos[P$chr == vec_chr[1]] - minimum
P$colours_chr[P$chr == vec_chr[1]] = vec_col[1]
maximum = max(P$consecutive_pos[P$chr == vec_chr[1]], na.rm=TRUE)

for (i in 2:length(vec_chr)) {
    # i = 2
    minimum = min(P$pos[P$chr == vec_chr[i]], na.rm=TRUE)
    P$consecutive_pos[P$chr == vec_chr[i]] = (P$pos[P$chr == vec_chr[i]] - minimum) + maximum + 1
    P$colours_chr[P$chr == vec_chr[i]] = vec_col[i]
    maximum = max(P$consecutive_pos[P$chr == vec_chr[i]], na.rm=TRUE)
    print(paste0("maximum=", maximum))
}

table(P$chr)

svg("res/test.svg", width=20, height=14)
par(mfrow=c(2,1))
plot(x=P$consecutive_pos, y=abs(P$b), type='n')
for (chr in vec_chr) {
    P_sub = P[P$chr == chr, ]
    points(x=P_sub$consecutive_pos, y=abs(P_sub$b), pch=19, col=P_sub$colours_chr)
}
grid()

plot(x=P$consecutive_pos, y=P$lod, type='n')
for (chr in vec_chr) {
    P_sub = P[P$chr == chr, ]
    points(x=P_sub$consecutive_pos, y=P_sub$lod, pch=19, col=P_sub$colours_chr)
}
grid()
# qqplot(seq(0, 1, length=nrow(P)), P$pval); grid()
dev.off()

head(P[order(P$lod, decreasing=TRUE), ])

svg("res/test.svg", width=7, height=7)
plot(df_y$y_true, df_y$y_pred_mean, type="n", 
    xlab="Observed", ylab="Predicted",
    main="y~pos+G+E\n(OLS - better than penalised regression models)"); grid()
for (i in 1:nrow(df_y)) {
    points(x=df_y$y_true[i], y=df_y$y_pred_mean[i], pch=19)
    arrows(x0=df_y$y_true[i],
        x1=df_y$y_true[i],
        y0=df_y$y_pred_mean[i]-df_y$y_pred_sd[i],
        y1=df_y$y_pred_mean[i]+df_y$y_pred_sd[i],
        code=3,
        length=0.1,
        angle=90,
        col="blue")
}
mod = lm(y_pred_mean ~ y_true, data=df_y)
abline(mod, col="blue")
legend("bottomright", legend=c(paste0("n=", nrow(df_y)),
                               paste0("p=", length(unique(paste0(P$chr, "-", P$pos)))),
                               paste0("corr=", round(cor(df_y$y_true, df_y$y_pred_mean)*100), "%"),
                               paste0("rmse=", round(sqrt(mean((df_y$y_true-df_y$y_pred_mean)^2)), 2))))
dev.off()



```

![test](../res/test.svg)
